File: 0438-find-all-anagrams-in-a-string/0438-find-all-anagrams-in-a-string.cpp
class Solution {
public:
    vector<int> findAnagrams(string s, string p) {
        vector<int> ans;
        int i=0;
        int j=0;
        unordered_map<char,int> mp;
        for(auto x:p) mp[x]+=1;
        int distinct_count = mp.size();
        int k = p.length();
        int n = s.size();
        //Grow the window
        while(j<k-1) {
            if( mp.find(s[j])!=mp.end()) {
                //It exists
                mp[s[j]]-=1;
                if(mp[s[j]]==0) distinct_count-=1;
            }
            j++;
        }


        while(j<n) {
            //Process the window
            if(mp.find(s[j])!=mp.end()) {
                mp[s[j]]-=1;
                if(mp[s[j]]==0) distinct_count-=1;
            }
            if(distinct_count==0) ans.push_back(i);
            //Prepare to move the window
            if(mp.find(s[i])!=mp.end()) {
                mp[s[i]]+=1;
                if(mp[s[i]]==1) distinct_count+=1;
            }

            //Move the window
            i++;
            j++;
        }


    return ans;    
    }
    
};

File: 0438-find-all-anagrams-in-a-string/0438-find-all-anagrams-in-a-string.py
from collections import Counter
class Solution:
    def findAnagrams(self, s: str, p: str) -> List[int]:
        counter = Counter(p)
        n = len(s)
        ans = []
        left=0
        numDifferentChar=len(counter)
        print(numDifferentChar,counter)
        for right in range(n):
            if s[right] in counter:
                counter[s[right]]-=1
                if counter[s[right]]==0:
                    numDifferentChar-=1
                elif counter[s[right]]==-1:
                    numDifferentChar+=1

            if right-left+1==len(p): 
                if numDifferentChar==0:
                    ans.append(left)
                
                if s[left] in counter:
                    counter[s[left]]+=1
                    if counter[s[left]]==0:
                        numDifferentChar-=1
                    elif counter[s[left]]==1:
                        numDifferentChar+=1
                left+=1
        return ans

